tripy/tripy/flat_ir/flat_ir.py

```diff
@@ -186,20 +210,35 @@ def register_tensor_and_collect_ops(tensor, seen_tensors):
                     op is not None
                 ), f"Tensor: {tensor} has no producer set. Did you use the constructor instead of the `build()` function?"
 
-                for inp in op.inputs:
-                    if inp not in inputs:
-                        register_tensor_and_collect_ops(inp, seen_tensors)
-
-                ops.append(op)
+                # If a constant is already been declared in the flatIR, reuse that constant instead of redefining the constant.
+                if isinstance(op, ConstantOp):
+                    constant_key = self._get_constant_key(op)
+                    if constant_key in self.constant_map:
+                        # Reuse existing constant
+                        existing_tensor = self.constant_map[constant_key]
+                        self.tensor_replacements[tensor.name] = existing_tensor
+                    else:
+                        # New unique constant, add to map
+                        self.constant_map[constant_key] = op.outputs[0]
+                        new_ops.append(op)
+                else:
+                    for inp in op.inputs:
+                        if inp not in inputs:
+                            register_tensor_and_collect_ops(inp, seen_tensors)
+                    new_ops.append(op)
 
         for start_tensor in outputs:
             register_tensor_and_collect_ops(start_tensor, seen_tensors)
 
+        # Apply tensor replacements
+        for op in new_ops:
+            op.inputs = [self.tensor_replacements.get(inp.name, inp) for inp in op.inputs]
+
         # Rebind the ops to tensors from this FlatIR
-        for op in ops:
+        for op in new_ops:
             op.inputs = [self.register_tensor(inp) for inp in op.inputs]
```

Comment:
is registering a tensor twice a side effect free?

Reply:
Where are we registering twice?